common/dynamic_zone_base.cpp

```diff
@@ -137,34 +150,80 @@ uint32_t DynamicZoneBase::SaveToDatabase()
 	return 0;
 }
 
-void DynamicZoneBase::AddCharacter(uint32_t character_id)
+bool DynamicZoneBase::AddMember(const DynamicZoneMember& add_member)
 {
-	DynamicZoneMembersRepository::AddMember(GetDatabase(), m_id, character_id);
-	GetDatabase().AddClientToInstance(m_instance_id, character_id);
-	SendInstanceAddRemoveCharacter(character_id, false); // stops client kick timer
+	if (HasMember(add_member.id))
+	{
+		return false;
+	}
+
+	DynamicZoneMembersRepository::AddMember(GetDatabase(), m_id, add_member.id);
+	GetDatabase().AddClientToInstance(m_instance_id, add_member.id);
+
+	ProcessMemberAddRemove(add_member, false);
+	SendServerPacket(CreateServerMemberAddRemovePacket(add_member, false).get());
+
+	return true;
 }
 
-void DynamicZoneBase::RemoveCharacter(uint32_t character_id)
+bool DynamicZoneBase::RemoveMember(uint32_t character_id)
 {
-	DynamicZoneMembersRepository::RemoveMember(GetDatabase(), m_id, character_id);
-	GetDatabase().RemoveClientFromInstance(m_instance_id, character_id);
-	SendInstanceAddRemoveCharacter(character_id, true); // start client kick timer
+	auto remove_member = GetMemberData(character_id);
+	return RemoveMember(remove_member);
 }
 
-void DynamicZoneBase::RemoveAllCharacters(bool enable_removal_timers)
+bool DynamicZoneBase::RemoveMember(const std::string& character_name)
 {
-	if (GetInstanceID() == 0)
+	auto remove_member = GetMemberData(character_name);
+	return RemoveMember(remove_member);
+}
+
+bool DynamicZoneBase::RemoveMember(const DynamicZoneMember& remove_member)
+{
+	if (remove_member.id == 0)
 	{
-		return;
+		return false;
 	}
 
-	if (enable_removal_timers)
+	DynamicZoneMembersRepository::RemoveMember(GetDatabase(), m_id, remove_member.id);
+	GetDatabase().RemoveClientFromInstance(m_instance_id, remove_member.id);
+
+	ProcessMemberAddRemove(remove_member, true);
+	SendServerPacket(CreateServerMemberAddRemovePacket(remove_member, true).get());
+
+	return true;
+}
+
+bool DynamicZoneBase::SwapMember(
+	const DynamicZoneMember& add_member, const std::string& remove_char_name)
+{
+	auto remove_member = GetMemberData(remove_char_name);
+	if (!add_member.IsValid() || !remove_member.IsValid())
 	{
-		SendInstanceRemoveAllCharacters();
+		return false;
 	}
 
+	// make remove and add atomic to avoid racing with separate world messages
+	DynamicZoneMembersRepository::RemoveMember(GetDatabase(), m_id, remove_member.id);
```

Comment:
This is fine, it was copy/pasted when code was moved and the comment is just a reminder to use a single server message for the operation to avoid races